IndexOutOfBoundsException when attempting to reduce ill-kinded types

The following snippet is a variant of that causing #2771:

```Scala
trait A { type L[G[F[_],_],H[F[_],_]] }
trait B { type L[F[_],_] }
trait C { type M <: A }
trait D { type M >: B }

object Test {
  def test(x: C with D): Unit = {
    def foo(a: A, b: B)(z: a.L[b.L,b.L]) = z
    def bar(y: x.M, b: B) = foo(y, b)
    def baz(b: B) = bar(b, b)
    baz(new B { type L[F[_],X] = F[X] })(1)
  }
}
```

The underlying issue is likely the same, though the error message is different:

    exception occurred while compiling Test.scala
    Exception in thread "main" java.lang.IndexOutOfBoundsException: 1
        at scala.collection.LinearSeqOptimized$class.apply(LinearSeqOptimized.scala:65)
        at scala.collection.immutable.List.apply(List.scala:84)
        at dotty.tools.dotc.core.Substituters$class.substParams(Substituters.scala:220)
        ...

The error disappears when the last line (the call to `baz`) is commented out. See #2771 for a more in-depth discussion.

Here is another example, producing a stack overflow again (like for #2771). It uses the untyped SKI fixpoint combinator and absurd bounds to create a non-terminating term.

```Scala
trait A { type S[X[_] <: [_] => Any, Y[_]] <: [_] => Any; type I[_] }
trait B { type S[X[_],Y[_]]; type I[_] <: [_] => Any }
trait C { type M <: B }
trait D { type M >: A }

object Test {
  def test(x: C with D): Unit = {
    def foo(a: A, b: B)(z: a.S[b.I,a.I][b.S[a.I,a.I]]) = z
    def bar(a: A, y: x.M) = foo(a,y)
    def baz(a: A) = bar(a, a)
    baz(new A { type S[X[_] <: [_] => Any, Y[_]] = [Z] => X[Z][Y[Z]]; type I[X] = X })(1)
  }
}
```
Again, the error disappears when the call to `baz` is commented out.

Note the use of curried type operators. One can also write an uncurried version, resulting in yet another error:
```Scala
trait A { type S[X[_,_], Y[_],_]; type I[_] }
trait B { type S[X[_],Y[_]]; type I[_,_] }
trait C { type M <: B }
trait D { type M >: A }

object Test {
  def test(x: C with D): Unit = {
    def foo(a: A, b: B)(z: a.S[b.I,a.I,b.S[a.I,a.I]]) = z
    def bar(a: A, y: x.M) = foo(a,y)
    def baz(a: A) = bar(a, a)
    baz(new A { type S[X[_,_], Y[_], Z] = X[Z,Y[Z]]; type I[X] = X })(1)
  }
}
```
produces the error

    exception occurred while compiling SKIUncurried.scala
    Exception in thread "main" java.lang.AssertionError: assertion failed
        at scala.Predef$.assert(Predef.scala:156)
        at dotty.tools.dotc.core.Types$TypeAccumulator.foldArgs$1(Types.scala:3979)

Interestingly, the error does *not* disappear in this case when call to `baz` is commented out.

Note that none of these examples uses recursion (neither on the term level nor on the type level). All the trouble is caused by bad bounds.

--- a/compiler/src/dotty/tools/dotc/core/TypeApplications.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeApplications.scala
@@ -343,11 +343,15 @@ class TypeApplications(val self: Type) extends AnyVal {
               }
             }
             if ((dealiased eq stripped) || followAlias)
-              try {
+              try
                 val instantiated = dealiased.instantiate(args)
                 if (followAlias) instantiated.normalized else instantiated
-              }
-              catch { case ex: IndexOutOfBoundsException => AppliedType(self, args) }
+              catch
+                case ex: IndexOutOfBoundsException =>
+                  AppliedType(self, args)
+                case ex: Throwable =>
+                  handleRecursive("try to instantiate", i"$dealiased[$args%, %]", ex)
+
             else AppliedType(self, args)
           }
           else dealiased.resType match {

--- a/tests/neg/i2887b.check
+++ b/tests/neg/i2887b.check
@@ -1,0 +1,30 @@
+-- Error: tests/neg/i2887b.scala:11:7 ----------------------------------------------------------------------------------
+11 |    baz(new A { type S[X[_] <: [_] =>> Any, Y[_]] = [Z] =>> X[Z][Y[Z]]; type I[X] = X })(1) // error
+   |       ^
+   |       Recursion limit exceeded.
+   |       Maybe there is an illegal cyclic reference?
+   |       If that's not the case, you could also try to increase the stacksize using the -Xss JVM option.
+   |       A recurring operation is (inner to outer):
+   |
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         ...
+   |
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]
+   |         try to instantiate Z[Z]

--- a/tests/neg/i2887b.scala
+++ b/tests/neg/i2887b.scala
@@ -1,13 +1,13 @@
-trait A { type S[X[_] <: [_] => Any, Y[_]] <: [_] => Any; type I[_] } // error // error
-trait B { type S[X[_],Y[_]]; type I[_] <: [_] => Any } // error
+trait A { type S[X[_] <: [_] =>> Any, Y[_]] <: [_] =>> Any; type I[_] }
+trait B { type S[X[_],Y[_]]; type I[_] <: [_] =>> Any }
 trait C { type M <: B }
 trait D { type M >: A }
 
 object Test {
   def test(x: C with D): Unit = {
-    def foo(a: A, b: B)(z: a.S[b.I,a.I][b.S[a.I,a.I]]) = z // error
+    def foo(a: A, b: B)(z: a.S[b.I,a.I][b.S[a.I,a.I]]) = z
     def bar(a: A, y: x.M) = foo(a,y)
     def baz(a: A) = bar(a, a)
-    baz(new A { type S[X[_] <: [_] => Any, Y[_]] = [Z] => X[Z][Y[Z]]; type I[X] = X })(1) // error // error
+    baz(new A { type S[X[_] <: [_] =>> Any, Y[_]] = [Z] =>> X[Z][Y[Z]]; type I[X] = X })(1) // error
   }
 }

